README.md
# maps_curved_line

[![pub package](https://img.shields.io/pub/v/maps_curved_line.svg)](https://pub.dartlang.org/packages/maps_curved_line)

A package to help draw curved lines on [Google Maps](https://pub.dev/packages/google_maps_flutter).
Generates a list of LatLng which can be used to render a curved polyline between the given
start and end LatLng.

<p>
  <img src="https://user-images.githubusercontent.com/8823044/78174319-96853100-7448-11ea-9389-fdd0dd12579b.png" width="290" />
  <img src="https://user-images.githubusercontent.com/8823044/78174332-9ab14e80-7448-11ea-80c6-0d2b825d853a.png" width="290" />
</p>

## Installation

First, add `maps_curved_lines` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

## Usage:

``` dart
import 'package:maps_curved_line/maps_curved_line.dart';
.
.

class _MapsPageState extends State<MapsPage> {

  final Set<Polyline> _polylines = Set();

  final LatLng _point1 = LatLng(12.947437, 77.681345);
  final LatLng _point2 = LatLng(12.948767, 77.689120);

  @override
  Widget build(BuildContext context) {
    _polylines.add(
        Polyline(
            polylineId: PolylineId("line 1"),
            visible: true,
            width: 2,
            //latlng is List<LatLng>
            patterns: [PatternItem.dash(30), PatternItem.gap(10)],
            points: MapsCurvedLines.getPointsOnCurve(_point1, _point2), // Invoke lib to get curved line points
            color: Colors.blue,
        )
    );

    return new Scaffold(
      body: GoogleMap(
        // Configure google maps widget are required
        .
        polylines: _polylines, // Add constructed polyline for curved line
        .
        .
      ),
    );
  }
}
```

CHANGELOG.md
## [1.0.2]
Update homepage

## [1.0.1]
Migrate to null safety and upgrade dependencies  (google_maps_flutter - v2.0.3)

## [0.2.0]
Upgrade dependencies (google_maps_flutter - v1.0.2) and flutter 1.22.1

## [0.1.1]
Add public screenshots to README

## [0.1.0]
Fix iOS example code and mark as stable. Add screenshots to README

## [0.0.2]
Improve package score by fixing hints and adding more description.

## [0.0.1]
Initial Release

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:maps_curved_lines_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/maps_curved_line.dart
library maps_curved_lines;

import 'dart:math';

import 'package:google_maps_flutter/google_maps_flutter.dart';

import 'maps/spherical_utils.dart';

class MapsCurvedLines {
  static List<LatLng> getPointsOnCurve(
      LatLng startLocation, LatLng endLocation) {
    // https://medium.com/better-programming/curved-lines-on-google-maps-2938bbb15f6a
    List<LatLng> path = [];
    double angle = pi / 2;
    // ignore: non_constant_identifier_names
    double SE =
        SphericalUtils.computeDistanceBetween(startLocation, endLocation)
            as double;
    // ignore: non_constant_identifier_names
    double ME = SE / 2.0;
    double R = ME / sin(angle / 2);
    // ignore: non_constant_identifier_names
    double MO = R * cos(angle / 2);

    double heading =
        SphericalUtils.computeHeading(startLocation, endLocation) as double;
    LatLng mCoordinate =
        SphericalUtils.computeOffset(startLocation, ME, heading);

    double direction =
        (startLocation.longitude - endLocation.longitude > 0) ? -1.0 : 1.0;
    double angleFromCenter = 90.0 * direction;
    LatLng oCoordinate = SphericalUtils.computeOffset(
        mCoordinate, MO, heading + angleFromCenter);

    path.add(endLocation);

    int num = 100;

    double initialHeading =
        SphericalUtils.computeHeading(oCoordinate, endLocation) as double;
    double degree = (180.0 * angle) / pi;

    for (int i = 1; i <= num; i++) {
      double step = i.toDouble() * (degree / num.toDouble());
      double heading = (-1.0) * direction;
      LatLng pointOnCurvedLine = SphericalUtils.computeOffset(
          oCoordinate, R, initialHeading + heading * step);
      path.add(pointOnCurvedLine);
    }

    path.add(startLocation);

    return path;
  }
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:maps_curved_line/maps_curved_line.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Maps Curved Lines Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MapsPage(),
    );
  }
}

class MapsPage extends StatefulWidget {
  @override
  _MapsPageState createState() => _MapsPageState();
}

class _MapsPageState extends State<MapsPage> {
  Completer<GoogleMapController> _controller = Completer();

  final Set<Polyline> _polylines = Set();
  final Set<Marker> _markers = Set();

  static final CameraPosition _initialPosition = CameraPosition(
    target: LatLng(12.947437, 77.685345),
    zoom: 15,
  );

  final LatLng _point1 = LatLng(12.947437, 77.681345);
  final LatLng _point2 = LatLng(12.948767, 77.689120);
  final LatLng _point3 = LatLng(12.947216, 77.680614);
  final LatLng _point4 = LatLng(12.948767, 77.609120);

  void prepareMarkers() {
    _markers.add(Marker(
      // This marker id can be anything that uniquely identifies each marker.
      markerId: MarkerId(_point1.toString()),
      //_lastMapPosition is any coordinate which should be your default
      //position when map opens up
      position: _point1,
      icon: BitmapDescriptor.defaultMarker,
    ));

    _markers.add(Marker(
      // This marker id can be anything that uniquely identifies each marker.
      markerId: MarkerId(_point2.toString()),
      //_lastMapPosition is any coordinate which should be your default
      //position when map opens up
      position: _point2,
      icon: BitmapDescriptor.defaultMarker,
    ));

    _markers.add(Marker(
      // This marker id can be anything that uniquely identifies each marker.
      markerId: MarkerId(_point3.toString()),
      //_lastMapPosition is any coordinate which should be your default
      //position when map opens up
      position: _point3,
      icon: BitmapDescriptor.defaultMarker,
    ));

    _markers.add(Marker(
      // This marker id can be anything that uniquely identifies each marker.
      markerId: MarkerId(_point4.toString()),
      //_lastMapPosition is any coordinate which should be your default
      //position when map opens up
      position: _point4,
      icon: BitmapDescriptor.defaultMarker,
    ));
  }

  void prepareCurvedPolylines() {
    _polylines.add(
        Polyline(
          polylineId: PolylineId("line 1"),
          visible: true,
          width: 2,
          //latlng is List<LatLng>
          patterns: [PatternItem.dash(30), PatternItem.gap(10)],
          points: MapsCurvedLines.getPointsOnCurve(_point1, _point2),
          color: Colors.blue,
        )
    );

    _polylines.add(
        Polyline(
          polylineId: PolylineId("line 2"),
          visible: true,
          width: 2,
          //latlng is List<LatLng>
          patterns: [PatternItem.dash(30), PatternItem.gap(10)],
          points: MapsCurvedLines.getPointsOnCurve(_point3, _point4),
          color: Colors.grey,
        )
    );
  }

  @override
  Widget build(BuildContext context) {

    prepareMarkers();
    prepareCurvedPolylines();

    return new Scaffold(
      body: GoogleMap(
        mapType: MapType.normal,
        initialCameraPosition: _initialPosition,
        polylines: _polylines,
        markers: _markers,
        onMapCreated: (GoogleMapController controller) {
          _controller.complete(controller);
        },
      ),
    );
  }
}
